test/integration/consul-container/test/upgrade/l7_traffic_management/resolver_subset_redirect_test.go

```diff
@@ -0,0 +1,234 @@
+		require.NoError(t, agentServices.server2ProxyV2.Restart())
+
+		// validate static-client is up and running
+		libassert.AssertContainerState(t, agentServices.client, "running")
+		libassert.HTTPServiceEchoes(t, "localhost", port, "")
+
+		// assert 3 static-server instances are healthy
+		libassert.AssertServiceHasHealthyInstances(t, node, libservice.StaticServer2ServiceName, false, 3)
+
+		// static-client upstream should have 1 healthy endpoint for v2.static-server-2
+		libassert.AssertUpstreamEndpointStatus(t, adminPort, "v2.static-server-2.default", "HEALTHY", 1)
+
+		// static-client upstream should connect to static-server-v2 because the default subset value is to v2 set in the service resolver
+		libassert.AssertFortioName(t, fmt.Sprintf("http://localhost:%d", port), "static-server-2-v2")
+
+		// validate static-client proxy admin is up
+		_, clientStatusCode, err := libassert.GetEnvoyOutput(adminPort, "stats", responseFormat)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, clientStatusCode, fmt.Sprintf("service cannot be reached %v", clientStatusCode))
+
+		// validate static-server proxy admin is up
+		_, serverStatusCode, err := libassert.GetEnvoyOutput(serverAdminPort, "stats", responseFormat)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, serverStatusCode, fmt.Sprintf("service cannot be reached %v", serverStatusCode))
+
+		// validate static-server-2 proxy admin is up
+		_, server2StatusCode, err := libassert.GetEnvoyOutput(server2AdminPort, "stats", responseFormat)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, server2StatusCode, fmt.Sprintf("service cannot be reached %v", server2StatusCode))
+
+		// validate static-server-2-v1 proxy admin is up
+		_, server2StatusCodeV1, err := libassert.GetEnvoyOutput(server2AdminPortV1, "stats", responseFormat)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, server2StatusCodeV1, fmt.Sprintf("service cannot be reached %v", server2StatusCodeV1))
+
+		// validate static-server-2-v2 proxy admin is up
+		_, server2StatusCodeV2, err := libassert.GetEnvoyOutput(server2AdminPortV2, "stats", responseFormat)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, server2StatusCodeV2, fmt.Sprintf("service cannot be reached %v", server2StatusCodeV2))
+
+		// certs are valid
+		libassert.AssertEnvoyPresentsCertURI(t, adminPort, libservice.StaticClientServiceName)
+		libassert.AssertEnvoyPresentsCertURI(t, serverAdminPort, libservice.StaticServerServiceName)
+		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPort, libservice.StaticServer2ServiceName)
+		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPortV1, libservice.StaticServer2ServiceName)
+		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPortV2, libservice.StaticServer2ServiceName)
+	}
+
+	for _, tc := range tcs {
+		t.Run(fmt.Sprintf("upgrade from %s to %s", tc.oldversion, tc.targetVersion),
+			func(t *testing.T) {
+				run(t, tc)
+			})
+	}
+}
+
+type Services struct {
```

Comment:
Can we return closure functions to run the assertions and proxy container restarts? We probably only need to return the client proxy

```go
func setupServiceAndSubsets(t *testing.T, cluster *libcluster.Cluster) (*libservice.Service, func(), func()){

    assertionFn := func() {
		_, serverAdminPort := agentServices.serverProxy.GetAdminAddr()
		_, server2AdminPort := agentServices.server2Proxy.GetAdminAddr()
		_, server2AdminPortV1 := agentServices.server2ProxyV1.GetAdminAddr()
		_, server2AdminPortV2 := agentServices.server2ProxyV2.GetAdminAddr()

              	_, server2StatusCodeV2, err := libassert.GetEnvoyOutput(server2AdminPortV2, "stats", responseFormat)
		require.NoError(t, err)
		assert.Equal(t, http.StatusOK, server2StatusCodeV2, fmt.Sprintf("service cannot be reached %v", server2StatusCodeV2))

		// certs are valid
		libassert.AssertEnvoyPresentsCertURI(t, adminPort, libservice.StaticClientServiceName)
		libassert.AssertEnvoyPresentsCertURI(t, serverAdminPort, libservice.StaticServerServiceName)
		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPort, libservice.StaticServer2ServiceName)
		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPortV1, libservice.StaticServer2ServiceName)
		libassert.AssertEnvoyPresentsCertURI(t, server2AdminPortV2, libservice.StaticServer2ServiceName)
    }

    proxyRestartFn := {
        require.NoError(t, agentServices.client.Restart())
        // restart other proxy containers
    }
    
    return  clientConnectProxy, assertionFn, proxyRestartFn
}
```

This way, all assertions are grouped together and the main test function makes calls to these returned functions before and after upgrade - reducing duplications.

Reply:
I like it, I'll make the changes now :)